Add align_double and align_int64 to `ocamlc -config`

toots opened the pull request:
These two values, particularly the `double` one, can be very useful to make low-level optimization of C code handling e.g. `floatarray`.

toots committed:
Add align_double and align_int64 to `ocamlc -config`

nojb commented:
> These two values, particularly the `double` one, can be very useful to make low-level optimization of C code handling e.g. `floatarray`.

Just to confirm, this is only an issue in 32-bits, do you agree? In 64-bits everything is 64-bit aligned anyway, right?

toots commented:
> > These two values, particularly the `double` one, can be very useful to make low-level optimization of C code handling e.g. `floatarray`.
> 
> Just to confirm, this is only an issue in 32-bits, do you agree? In 64-bits everything is 64-bit aligned anyway, right?

Maybe? Looks like the `configure` test has at least 3 different branches. I think, from a user perspective, it makes sense to have a clear data point without having to dig into specifics of the underlying architecture.

shindere commented:
Overall the code looks okay to me.

I have one remark and one question.

Question: can you confirm that the values in `utils/config.fixed.ml`
really are what you defined, i.e. true?

Remark: in the documentation strings in `utils/config.mli`, there are
two occurrences of `Wether the compiler and runtime needs to ...` where
`needs` should be `need`.

xavierleroy commented:
> These two values, particularly the double one, can be very useful to make low-level optimization of C code handling e.g. floatarray.

Agreed.  But they are already readily accessible from C code: just include `<caml/m.h>` and test for `ARCH_ALIGN_DOUBLE` and `ARCH_ALIGN_INT64`.  Why do you also need this information in the output of `ocamlc -config` ?

toots commented:
> > These two values, particularly the double one, can be very useful to make low-level optimization of C code handling e.g. floatarray.
> 
> Agreed. But they are already readily accessible from C code: just include `<caml/m.h>` and test for `ARCH_ALIGN_DOUBLE` and `ARCH_ALIGN_INT64`. Why do you also need this information in the output of `ocamlc -config` ?

`ocamlc -config` is pretty convenient to use with the existing tooling such as `dune` to have conditional compilation at build system level. 

`ocamlc -config` values are exported durectly as `dune` variables. (I would add a link but it looks like the dune doc is broken at the moment).

Otherwise, one has to write a test C program that does the include/printf dance and execute it to get its result.

Also, I was pretty surprised to see an unscoped global macro named like that. `ARCH_ALIGN_DOUBLE` does not seem OCaml specific and could potentially clash with other publicly defined macros. Not sure if it's a good idea to use it this way, it should also be renamed and named-spaced in my opinion.

toots commented:
> Overall the code looks okay to me. I have one remark and one question. Question: can you confirm that the values in `utils/config.fixed.ml` really are what you defined, i.e. true? Remark: in the documentation strings in `utils/config.mli`, there are two occurrences of `Wether the compiler and runtime needs to ...` where `needs` should be `need`.

Thanks! It's carnival time here in New Orleans but I plan on addressing these comments as soon as reality is back to normal! 🎭

xavierleroy commented:
> ocamlc -config is pretty convenient to use with the existing tooling such as dune to have conditional compilation at build system level.

Point well taken.  I'm still curious why you need to know the alignment constraints of the target machine to conditionally select things at build system level.  My experience so far is that you need to `#ifdef` on these constraints within C code, and nowhere else.

> Also, I was pretty surprised to see an unscoped global macro named like that. ARCH_ALIGN_DOUBLE does not seem OCaml specific and could potentially clash with other publicly defined macros. Not sure if it's a good idea to use it this way, it should also be renamed and named-spaced in my opinion.

First, you're changing the topic.  Second, I don't agree with your other suggestion.  Like most C libraries, OCaml tries to make function names and global variable names unique by prefixing them with `caml_`.  That's a necessity because all those names are visible at link-time.

But like most C libraries, OCaml doesn't need to make and macro names unique, because they are visible only if you `#include` the corresponding header, and in the rare case you get a conflict with another header, you can always `#undef` the problematic names.

For example, `<X11/Xlib.h>` proudly defines macros `Bool`, `True` and `False`, and while it did break some of my code in the past, it is considered acceptable.  OCaml's header files don't get anywhere this amount of annoyance :-)

toots commented:
> Overall the code looks okay to me. I have one remark and one question. Question: can you confirm that the values in `utils/config.fixed.ml` really are what you defined, i.e. true?

Good call, thanks. `boot/ocamlc` has been updated in the PR and now:
```
% ./boot/ocamlrun ./boot/ocamlc -config
...
flat_float_array: true
align_double: true
align_int64: true
```

> Remark: in the documentation strings in `utils/config.mli`, there are two occurrences of `Wether the compiler and runtime needs to ...` where `needs` should be `need`.

Fixed too, thanks for pointing out!

toots commented:
> > ocamlc -config is pretty convenient to use with the existing tooling such as dune to have conditional compilation at build system level.
> 
> Point well taken. I'm still curious why you need to know the alignment constraints of the target machine to conditionally select things at build system level. My experience so far is that you need to `#ifdef` on these constraints within C code, and nowhere else.
> 
> > Also, I was pretty surprised to see an unscoped global macro named like that. ARCH_ALIGN_DOUBLE does not seem OCaml specific and could potentially clash with other publicly defined macros. Not sure if it's a good idea to use it this way, it should also be renamed and named-spaced in my opinion.
> 
> First, you're changing the topic. Second, I don't agree with your other suggestion. Like most C libraries, OCaml tries to make function names and global variable names unique by prefixing them with `caml_`. That's a necessity because all those names are visible at link-time.
> 
> But like most C libraries, OCaml doesn't need to make and macro names unique, because they are visible only if you `#include` the corresponding header, and in the rare case you get a conflict with another header, you can always `#undef` the problematic names.
> 
> For example, `<X11/Xlib.h>` proudly defines macros `Bool`, `True` and `False`, and while it did break some of my code in the past, it is considered acceptable. OCaml's header files don't get anywhere this amount of annoyance :-)

Fair enough! As a developer I don't really feel confortable using a macro with a generic name but, as you pointed out, this is done some other places.

If I have this value explicitly available through `dune`/`ocamlc -config`, I would personally use it to generate a custom `C` macro and be happy forever after 🙂

shindere commented:
Thanks a lot. I am notsure that the PR itself does need to update
`boot/ocamlc\`? Can we perhaps wait the next bootstrap, or do we have to
have correct values immediately?

Actually my question was unclear, sorry about that. What I was asking
was whether  the values for hte fixed configuration were really intended
to be true, I wanted to make sure you had asked yourself about that. I
was not trying to ensure that you deed a bootstrap to update the boot
compiler and I don't actually knowwhether this really needs to happen in
this PR.

toots commented:
> Thanks a lot. I am notsure that the PR itself does need to update `boot/ocamlc\`? Can we perhaps wait the next bootstrap, or do we have to have correct values immediately? Actually my question was unclear, sorry about that. What I was asking was whether the values for hte fixed configuration were really intended to be true, I wanted to make sure you had asked yourself about that. I was not trying to ensure that you deed a bootstrap to update the boot compiler and I don't actually knowwhether this really needs to happen in this PR.

Ha I see. I'm not sure what the values should be in fact, the doc says:
> (* Configuration for the boot compiler. The compiler should refuse to bootstrap
   if configured with values which would contradict the configuration below.
   The values below are picked to trigger errors if accidentally used in the
   compiler (e.g. for the C compiler). *)
   
 In this case, the values could be `true` or `false` depending on the architecture/build configuration. Seems like this file has to match the general configuration module but that the effective parts to catch errors are with the values for `c_compiler`, `native_pack_linker` etc., in which case the default values added in this PR and set in this file would not matter much correct?

shindere commented:
Perhaps it's best to listen to @dra27 on this?

dra27 commented on utils/config.mli:
```diff
@@ -223,6 +223,13 @@ val flat_float_array : bool
 (** Whether the compiler and runtime automagically flatten float
     arrays *)
 
+val align_double : bool
+(** Wether the compiler and runtime need to align double values.
+    If [false], a [floatarray] value can be casted into a C array of double. *)
+
+val align_int64 : bool
+(** Wether the compiler and runtime need to align int64 values *)
```
Language nits only:
```suggestion
val align_double : bool
(** Whether the compiler and runtime need to align double values.
    If [false], a [floatarray] value can be cast to a C array of doubles. *)

val align_int64 : bool
(** Whether the compiler and runtime need to align int64 values *)
```

dra27 reviewed (approved):
LGTM, apart from a couple of tweaks to the .mli file. The update to `config.fixed.ml` is indeed correct - the alignment requirements won't affect any of the _outputs_ of the compiler, and therefore you just pick a value in the bootstrap.

toots commented:
> LGTM, apart from a couple of tweaks to the .mli file. The update to `config.fixed.ml` is indeed correct - the alignment requirements won't affect any of the _outputs_ of the compiler, and therefore you just pick a value in the bootstrap.

Done, thanks!

toots commented:
Looks like I totally forgot to merge this. Just fixed the conflicts. Planning on merging unless someone is opposed to it.

gasche commented:
There is a change to boot/ocamlc in your PR, you need to redo this step now to not put back an older compiler version. Did you already redo the bootstrap?

toots commented:
> There is a change to boot/ocamlc in your PR, you need to redo this step now to not put back an older compiler version. Did you already redo the bootstrap?

Yes I did!

```
% ./boot/ocamlrun ./boot/ocamlc -config                                                                                                                  15:05:27
version: 5.4.0+dev0-2024-08-25
...
```

dra27 commented:
Bootstrap doesn't appear to be correct (just noting not to merge yet)

dra27 commented:
I can't repeat the bootstrap in cb379322914c3cf2715a3469604ceb6ba97fe888 (`boot/ocamllex` differs) or 12ddfed305ea2754d26766d738d6adefe6d8e92b (both boot artefacts differ)

dra27 commented:
Rebased to clean out the edits in merge commits - this shouldn't require a bootstrap. Should be good to go now, thanks!

toots commented:
Thanks so much for the help!

Octachron commented:
It seems that the lack of handling for 32bit architecture is making the compiler fails due to a syntax error on 32 bits architecture.

dra27 commented:
Is there a pointer to where the problem is, @Octachron? Apart from the fact the i386 test here was fine, I'm struggling to see by inspection how any path could lead to a syntax error here?

Octachron commented:
Sorry, this is indeed ARM32 specific, for which the generated `config.generated.ml` file ends up containing:

```ocaml
let align_double = true,
let align_int64 = true,
```

shindere commented:
Working on a fix. See my branch fix-12019 on shindere/ocaml. It's
currently going through prechecki:

https://ci.inria.fr/ocaml/job/precheck/984

dra27 commented:
Thanks! @mtelvers had spotted it on our arm32 builds as well and we were about to open a PR with the same fix!
